src/openms/source/QC/QCBase.cpp

```diff
@@ -37,4 +37,46 @@
 namespace OpenMS
 {
   const std::string QCBase::names_of_requires[] = {"fail", "raw.mzML", "postFDR.featureXML", "preFDR.featureXML", "contaminants.fasta", "trafoAlign.trafoXML"};
+
+  QCBase::SpectraMap::SpectraMap() = default;
+
+  QCBase::SpectraMap::SpectraMap(const MSExperiment& exp)
+  {
+    calculateMap(exp);
+  }
+
+  QCBase::SpectraMap::~SpectraMap() = default;
+
+  void QCBase::SpectraMap::calculateMap(const MSExperiment& exp)
+  {
+    map_to_index_.clear();
+    for (Size i = 0; i < exp.size(); ++i)
+    {
+      map_to_index_[exp[i].getNativeID()] = i;
+    }
+  }
+
+  const UInt64& QCBase::SpectraMap::at(const String& identifier) const
+  {
+    if (map_to_index_.find(identifier) == map_to_index_.end())
+    {
+      throw Exception::InvalidParameter(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "No spectrum with given identifier in MSExperiment!");
+    }
+    return map_to_index_.at(identifier);
```

Comment:
this causes a second traversal of the rb-tree. to avoid that, simply store the pointer obtained via find() earlier and return it->second